syntax = "proto2";
option optimize_for = SPEED;
package zlog_ceph_proto;

// metadata for objects linking to head object
message LinkObjectHeader {
  required string hoid = 1;
}

// metadata for head objects
message HeadObjectHeader {
  required bool deleted = 1;
  required string prefix = 2;
  optional uint64 max_epoch = 3;
}

// head object methods and replies
message CreateView {
  required uint64 epoch = 1;
  required bytes data = 2;
}

message ReadView {
  required uint64 epoch = 1;
  optional uint32 max_views = 2 [default = 100];
}

message View {
  required uint64 epoch = 1;
  required bytes data = 2;
}

message Views {
  repeated View views = 1;
}

// metadata for objects storing log entries
message LogObjectHeader {
  optional uint64 epoch = 1;
  optional uint64 max_pos = 2;
}

// log entry metadata

message LogEntry {
  enum Flags {
    INVALID = 1;
    FORCED  = 2;
  }
  optional uint32 flags = 1 [default = 0];
  optional uint32 offset = 2;
  optional uint32 length = 3;
}

// log object methods and replies

message ReadMaxPos {
  required uint64 epoch = 1;
}

message MaxPos {
  optional uint64 pos = 1;
}

message Seal {
  required uint64 epoch = 1;
}

message WriteEntry {
  required uint64 epoch = 1;
  required uint64 pos = 2;
  required bytes data = 3;
}

message ReadEntry {
  required uint64 epoch = 1;
  required uint64 pos = 2;
}

message InvalidateEntry {
  required uint64 epoch = 1;
  required uint64 pos = 2;
  required bool force = 3;
}
